From: Herman van Hazendonk <github.com@herrie.org>
Subject: [PATCH] language plugins: link Qt6Core5Compat for QTextCodec

None of the 16 language plugins could be loaded, so MaliitServer crashed on
startup once it got as far as creating the input method:

    MaliitKeyboard::Logic::WordEngine::setSpellcheckerEnabled(bool)
    InputMethod::InputMethod(MAbstractInputMethodHost*)
    MaliitKeyboardPlugin::createInputMethod(MAbstractInputMethodHost*)
    MIMPluginManagerPrivate::loadPlugin(QDir const&, QString const&)

WordEnginePrivate::loadPlugin() only assigns languagePlugin inside
"if (plugin)", so when QPluginLoader::instance() returns nullptr the pointer
stays null and the next call dereferences it.

The plugins fail to load because they have an undefined reference to
QTextCodec::codecForName(). plugins/westernsupport uses QTextCodec (via
QtCore5Compat) in spellchecker.cpp and correctly declares "QT += core5compat",
but westernsupport is CONFIG += staticlib, so that only applies to its own
compilation. Each language plugin links libwesternsupport.a into its own shared
object and therefore has to link Qt6Core5Compat itself, which none of them did:

    $ readelf -d libenglishplugin.so | grep NEEDED
     ... libpresage.so.1, libhunspell-1.7.so.0, libQt6Core.so.6, libstdc++.so.6 ...

Add core5compat to every plugin that links westernsupport. Qt6Core5Compat is
already on the image, so only the link was missing.

Upstream-Status: Pending
Signed-off-by: Herman van Hazendonk <github.com@herrie.org>
---
--- a/plugins/ar/src/src.pro
+++ b/plugins/ar/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/cs/src/src.pro
+++ b/plugins/cs/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/da/src/src.pro
+++ b/plugins/da/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/de/src/src.pro
+++ b/plugins/de/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/en/src/src.pro
+++ b/plugins/en/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/es/src/src.pro
+++ b/plugins/es/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/fi/src/src.pro
+++ b/plugins/fi/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/fr/src/src.pro
+++ b/plugins/fr/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/he/src/src.pro
+++ b/plugins/he/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/hu/src/src.pro
+++ b/plugins/hu/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/it/src/src.pro
+++ b/plugins/it/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/nl/src/src.pro
+++ b/plugins/nl/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/pl/src/src.pro
+++ b/plugins/pl/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/pt/src/src.pro
+++ b/plugins/pt/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/ru/src/src.pro
+++ b/plugins/ru/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
--- a/plugins/sv/src/src.pro
+++ b/plugins/sv/src/src.pro
@@ -6,6 +6,7 @@
 TEMPLATE        = lib
 CONFIG         += plugin
 QT             += widgets
+QT             += core5compat
 INCLUDEPATH    += \
     $${TOP_SRCDIR}/src/ \
     $${TOP_SRCDIR}/src/lib/ \
